### PR TITLE
Travel Camera

### PR DESCRIPTION
**Pull request by SlamBamActionman**
## About the PR

Adds the Travel Camera to the game.
While this was primarily made to support the Tourist role space-wizards/docs#129, this change also adds the Travel Camera to maintenance as rare loot, and to the Syndicate uplink.

The Travel Camera is functionally similar to a Flash, though with one key difference: When used in target mode (i.e. Harm Mode + LeftClick, not AoS) the Travel Camera knocks the target down, similar to the effect seen when a Head Revolutionary converts someone. This is a 3 second stun (same as a conversion stun / banana peel slip) but can be avoided with the same methods used to resist normal flashes. 

- The Travel Camera stun can NOT be used to convert people to the Revolution.
- The Travel Camera has 4 charges, and can't be recharged.
- Available in the Uplink for 2 TC.

In a future PR, it could be possible to have the camera print out a photo of the area where it was taken. That's beyond the scope of this PR however.

## Why / Balance

The Travel Camera adds a layer of deniability for Head Revolutionaries to prevent metagaming, as hearing a flashing noise away from Security is no longer a guaranteed way of ascertaining a Revolution is underway. It is also an okay offensive/defensive option that allows jobs normally resistant to slips (janitor, magboots) to be stunned, but the limited stun duration and amount keeps it from being oppressive.

The two most common ways the Travel Camera will be seen is via the non-antag Tourist role or as a Syndicate uplink item. It is also available as very rare maintenance loot, to give normal crew access and ensure it's not an item that makes you valid. 

As a Syndicate item, it's not a great offensive purchase for its cost (compare to Syndicate soap, which stuns for 5 seconds). This is intentional; it makes the item a much more favorable choice when disguised as a Tourist, or if a Kill objective target is an aforementioned slip-resistant role. Unlike soap you can also not avoid the effect (by walking). The Travel Camera can also be used to cause panic for the crew that a Revolution might be underway.

## Technical details

Adds the `StunTime` property for the `FlashComponent`, which when set enables the item to stun the target. This does not work when used in AoE flash mode.
The `RevolutionaryRuleSystem` also now makes sure to check that it's a Flash being used when converting, via an additional property in `HeadRevolutionaryComponent`

## Media

Video:
[video link](https://www.youtube.com/watch?v=6XHjSekRn6E)

Image:
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## To be done
- ~~Make sure only Flashes can be used to convert Revolutionaries~~ Done!
- ~~Change stun duration values~~ Done!
- ~~Add to uplink~~ Done!
- ~~Create sprites~~ Done!

**Changelog**

:cl:
- add: Added the Travel Camera, a flash item available as maintenance loot and via Uplink.

### CODE REVIEW

**Comment by asperger-sind** [comment]
The uplink camera looks like a sidegrade on the extra bright (flashing) lantern, you should probably look into that if you didn't already.

**Comment by SlamBamActionman** [comment]
> The uplink camera looks like a sidegrade on the extra bright (flashing) lantern, you should probably look into that if you didn't already.

That's a fair point, I actually missed this! Comparing the two I think that they fulfill different purposes; while you *can* use the travel camera as an AoE flash, due to the limited number of uses you probably want to utilize it in the target mode instead. Flash lamp thus works much better for AoE with its many uses.

I changed the stun time back to the Rev stun time (3 seconds), which is the same as a banana peel slip. This allows time to cuff a target, but the loud noise makes it hard to use covertly. Nocturine is better for that.

**Comment by SlamBamActionman** [comment]
Ended up removing the AoE completely:

a) Makes it easier to wear without accidentally using it (I didn't mention but of course you can equip the camera around your neck!)
c) Removes using a non-optimal feature of the item
b) Makes it more unique from the lantern
d) Will likely make it easier to implement and convey how to do the future Tourist "Photograph" objective

File: Content.Shared/Revolutionary/Components/HeadRevolutionaryComponent.cs [hunk 1]
```diff
@@ -21,4 +21,10 @@ public sealed partial class HeadRevolutionaryComponent : Component
     /// </summary>
     [DataField, ViewVariables(VVAccess.ReadWrite)]
     public TimeSpan StunTime = TimeSpan.FromSeconds(3);
+
+    /// <summary>
+    /// Which entity tags that indicate if an item can flash and convert another player.
+    /// </summary>
+    [DataField("tags")]
+    public string[] ConvertItemTags = { "ConvertItem" };
```
Comment (metalgearsloth) [inline Content.Shared/Revolutionary/Components/HeadRevolutionaryComponent.cs:29]: This should be using a whitelist.

File: Content.Server/Flash/FlashSystem.cs [hunk 2]
```diff
@@ -139,6 +140,11 @@ private bool UseFlash(EntityUid uid, FlashComponent comp, EntityUid user)
             _stun.TrySlowdown(target, TimeSpan.FromSeconds(flashDuration/1000f), true,
                 slowTo, slowTo);
 
+            if (stunTime != null)
+            {
+                _stun.TryParalyze(target, TimeSpan.FromSeconds(stunTime.Value/1000f), true);
+            }
+
```
Comment (metalgearsloth) [inline Content.Server/Flash/FlashSystem.cs:147]: Should this be a separate component instead? i.e. StunAfterFlashComponent then subscribe to AfterFlashEvent.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by UbaserB** [comment]
Cute

**Comment by SlamBamActionman** [comment]
In addition to some code review changes, maintainers provided feedback to the Camera on the Tourist design proposal PR space-wizards/docs#129. My proposed changes are the following: 

It now holds 10 charges, has no AoE flash (for balance and also to avoid triggering it accidentally when equipping), no longer trips the target.

Compare this to the Flash Lantern, which has 15 charges, AoE flash, but is also identifiable as contraband upon inspection. Essentially the Travel Camera trades 5 charges for making you not valid, and also serves as a good Tourist impersonation tool. It's also largely ineffective against Security due to their sunglasses, making it suboptimal for shitter behavior.

**Comment by 0x6273** [comment]
> In a future PR, it could be possible to have the camera print out a photo of the area where it was taken. That's beyond the scope of this PR however.

It should be in scope. A camera that can't take pictures doesn't really make sense, and has too much overlap with the flash lantern.

**Comment by SlamBamActionman** [comment]
> > In a future PR, it could be possible to have the camera print out a photo of the area where it was taken. That's beyond the scope of this PR however.
> 
> It should be in scope. A camera that can't take pictures doesn't really make sense, and has too much overlap with the flash lantern.

I agree that it is definitely a thing that the camera should do in the future and I am keen on having it implemented, but seeing as how it is primarily a tool for the proposed Tourist role to complete their objectives (especially now that Revs are removed and the flash interaction is no longer relevant), the camera being able to actually print photos is ultimately a nice-to-have from a gameplay perspective.

**Comment by VasilisThePikachu** [comment]
I dont think theres a point to this, even for head revs. We already had a pr to make a diffrent flash for revs (#22061, #23451) and they were both pretty much unwanted for revs.

Plus we no longer have revs now. And i dont know how i feel about giving tourist a flasher... as far as i know they aint supposed to get syndicate. So theres no point for them to get a flashing ability.

This is just a reskinned flasher. I would love to see it actually take photos instead.

**Comment by SlamBamActionman** [comment]
> I dont think theres a point to this, even for head revs. We already had a pr to make a diffrent flash for revs (#22061, #23451) and they were both pretty much unwanted for revs.

As listed in the PR starter post, the Travel Camera would _not_ be able to be used by Head Revs to convert and this PR includes code specifically to prevent that. It is not designed to be another flash option like linked PRs. The interaction with the Revs gamemode would've not come from the item being used to convert, but to counter the metagame of crew hearing a flash and automatically assuming it's Revs. 

Either way the point is moot now since Revs are gone.

> And i dont know how i feel about giving tourist a flasher... as far as i know they aint supposed to get syndicate. So theres no point for them to get a flashing ability. 

As listed in the Tourist design proposal https://github.com/space-wizards/docs/pull/129 the Tourist's use of a Travel Camera is to utilize it to complete some of their objectives. While I agree that on paper that the actual objective action is basically hitting someone with the camera object, including the flash component is an obvious and thematically fitting addition while also being straightforward to implement. I'd hate for the photography feature to be a blocker for implementing the Tourist role. 

The Tourist will naturally have some charges left even after completing their objectives; the flash of the camera can then be used defensively or for further roleplay opportunities. An item capable of flashing is a relatively harmless role spawn item, especially compared to Detective or Bartenders (not to mention the entire Science department get flashes for free).

> Neither do i think syndies would use it. "Blending in as a toursit" wont easily escape you sec

Once the Tourist is implemented the ability to disguise as them _should_ be an option for Syndicate agents; part of the Tourist's design is the inability for crew to know that someone is a Tourist with 100% certainty, and antags should be able to exploit this. Due to the Tourist's objectives it won't be weird to see them requesting access to areas or personnel on the station. 

Since the Tourist will spawn with the Travel Camera there must be alternate ways to get it, to ensure it's not being used as a way to verify the Tourist's legitimacy. Uplink and rare maintenance drop are the best ways to go about this.

**Comment by SlamBamActionman** [comment]
Again, I fully support the ability for the camera to take photos and it should be a feature of it down the line. Considering the primary usage of the Travel Camera extends beyond that I do not see an issue with merging it without the photo feature. 

I mean we have a surgery duffle bag in uplink, it's not like we're adverse to including things now that don't have the full functionality coded.

**Comment by SlamBamActionman** [comment]
This PR is being set to draft until the Tourist PR is accepted (likely after the upcoming antag system rework)

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Pull request closed without merging**